### PR TITLE
Added mx.stack c++ frontend impl

### PR DESCRIPTION
**Pull request by j-csc**
This pr will add `stack` support for `mlx.core` c++ frontend.

```cpp
// Functions:
array stack(const std::vector<array>& arrays, int axis, StreamOrDevice s = {});
array stack(const std::vector<array>& arrays, StreamOrDevice s = {});

// Usage:
auto x = array({1, 2, 3}, {3});
auto y = array({4, 5, 6}, {3});
auto z = std::vector<array>{x, y};

stack(z, axis=1) // shape (3,2)
```

Will add Python bindings in next PR. Also might need to work on making it a `Primitive` if needed @awni

### CODE REVIEW

File: mlx/utils.cpp [hunk 1]
```diff
@@ -49,6 +49,31 @@ std::vector<int> broadcast_shapes(
   return out_shape;
 }
 
+bool is_same_size_and_shape(const std::vector<array>& arrays) {
+  if (arrays.empty())
+    return true;
+
+  return std::all_of(arrays.begin() + 1, arrays.end(), [&](const array& a) {
+    return ((a.shape() == arrays[0].shape()) && a.dtype() == arrays[0].dtype());
```
Comment (angeloskath) [inline mlx/utils.cpp:57]: I would change the name to `is_same_shape` and remove the dtype check which isn't in the name either.

File: mlx/ops.cpp [hunk 2]
```diff
@@ -591,6 +591,29 @@ array concatenate(
...
+    throw std::invalid_argument("No arrays provided for stacking");
+  }
+  if (!is_same_size_and_shape(arrays)) {
+    throw std::invalid_argument("All arrays must have the same shape");
+  }
+  int normalized_axis = normalize_axis(axis, arrays[0].ndim() + 1);
+  std::vector<array> new_arrays;
+  new_arrays.reserve(arrays.size());
+  for (auto& a : arrays) {
+    new_arrays.emplace_back(expand_dims(a, normalized_axis, s));
```
[the remark this answers is not in the document]
Comment (j-csc) [reply mlx/ops.cpp:609]: made the change in concat, used `result_type`

**Review by angeloskath — CHANGES_REQUESTED** [review]
Thanks for the changes. It looks great. If it isn't too much trouble I will ask for a few more updates. I hope it 's ok.

- Could you rebase on top of main and run the style checks?
- Could you also implement the python bindings? It is quite a simple addition after the work you 've already done.

Thanks again!

**Comment by j-csc** [comment]
pushed up the fix, lmk if there's anything else

File: python/src/ops.cpp [hunk 3]
```diff
@@ -2150,6 +2150,38 @@ void init_ops(py::module_& m) {
...
+        }
+      },
+      "arrays"_a,
+      py::pos_only(),
+      "axis"_a = 0,
+      py::kw_only(),
+      "stream"_a = none,
+      R"pbdoc(
+      stack(arrays: List[array], axis: Optional[int] = 0, *, stream:
+      Union[None, Stream, Device] = None) -> array
```
[the remark this answers is not in the document]
Comment (j-csc) [reply python/src/ops.cpp:2171]: will get these fixed up

File: python/src/ops.cpp [hunk 4]
```diff
@@ -2150,6 +2150,38 @@ void init_ops(py::module_& m) {
...
+
+      Stacks the arrays along a new axis.
+
+      Args:
+          arrays (list(array)): A list of arrays to stack.
+          axis (int, optional): The axis in the result array along which the
+          input arrays are stacked. Defaults to 0. stream (Stream, optional):
```
Comment (angeloskath) [inline python/src/ops.cpp]: I am really sorry to have to do this to you again but could you fix the style here. Basically `input` should be indented and stream should be in the line below.

File: python/src/ops.cpp (hunk 3 above)
Comment (angeloskath) [inline python/src/ops.cpp]: Same as below style change, could you put it all in one line to match the rest of the signatures? Maybe in the future we 'll enforce all of them in 80 or 120 chars but for now...

File: python/src/ops.cpp [hunk 5]
```diff
@@ -2150,6 +2150,38 @@ void init_ops(py::module_& m) {
...
+          Stream or device. Defaults to None.
+
+      Returns:
+          array: Stacked array of the input arrays. Default stacked along
+          axis 0.
```
Comment (angeloskath) [inline python/src/ops.cpp]: Indent or even better simplify to `array: The resulting stacked array`

**Review by angeloskath — CHANGES_REQUESTED** [review]
Looks perfect. I left some style comments that currently break the docs if not updated.

Sorry for the back and forth. In the future if you want I could apply these minor changes myself and then merge. Just let me know if it is ok. I didn't do it by default cause I would show up as a co-author and steal a bit of your thunder which I didn't want.

**Comment by j-csc** [comment]
> Looks perfect. I left some style comments that currently break the docs if not updated.
> 
> Sorry for the back and forth. In the future if you want I could apply these minor changes myself and then merge. Just let me know if it is ok. I didn't do it by default cause I would show up as a co-author and steal a bit of your thunder which I didn't want.

Not at all, just happy to help! Will push up these fixes

**Comment by angeloskath** [comment]
@awni waiting on your approval on this.

**Review by awni — APPROVED** [review]

**Pull request merged**